Nested belongs-to validators gives error

Hey,

I've got nested `belongs-to` validators and the relationship does not have to exist. Eg.

```javascript
//model/foo.js
var Validations = buildValidations({
    bar: validator('belongs-to')
});

export default DS.Model.extend(Validations, {
    name: DS.attr(),
    bar: DS.belongsTo('bar')  
});

//model/bar.js
var Validations = buildValidations({
    baz: validator('belongs-to')
});

export default DS.Model.extend(Validations, {
    name: DS.attr(),
    baz: DS.belongsTo('baz')
});

//model/baz.js
var Validations = buildValidations({
    name: validator('presence', true)
});

export default DS.Model.extend(Validations, {
    name: DS.attr()  
});
```

Creating a `foo` model and calling validate without adding `bar` and `baz` now triggers and error.

```javascript

let model = this.store.createRecord('foo');
model.validate({}, true).then(({model, validations}) => {
    console.log(model, validations); 
});
``` 

gives

```
ember.debug.js:32127 Error: Assertion Failed: Cannot call get with 'validations' on an undefined object.
    at new Error (native)
    at Error.EmberError (http://localhost:4202/assets/vendor.js:26688:21)
    at assert (http://localhost:4202/assets/vendor.js:16447:13)
    at Object.assert (http://localhost:4202/assets/vendor.js:26452:34)
    at get (http://localhost:4202/assets/vendor.js:30911:22)
    at http://localhost:4202/assets/vendor.js:69404:20
    at tryCatch (http://localhost:4202/assets/vendor.js:63616:14)
    at invokeCallback (http://localhost:4202/assets/vendor.js:63631:15)
    at publish (http://localhost:4202/assets/vendor.js:63599:9)
    at http://localhost:4202/assets/vendor.js:42155:7
``` 

I've traced it to the `belongs-to` validator and temporarily added a check for the model variable:

```javascript

export default Base.extend({
  validate(value) {
    if (value) {
      if (canInvoke(value, 'then')) {
        return value.then(model => {
            return model && get(model, 'validations');
        });
      }
      return get(value, 'validations');
    }

    return true;
  }
});

```
Do you have any input on how to solve this? Is there any other way to define the nested `belongs-to`?

Cheers,
Carl

--- a/addon/validators/belongs-to.js
+++ b/addon/validators/belongs-to.js
@@ -83,7 +83,7 @@ export default Base.extend({
   validate(value) {
     if (value) {
       if (canInvoke(value, 'then')) {
-        return value.then(model => get(model, 'validations'));
+        return value.then(model => model ? get(model, 'validations') : true);
       }
       return get(value, 'validations');
     }

--- a/addon/validators/has-many.js
+++ b/addon/validators/has-many.js
@@ -59,7 +59,7 @@ export default Base.extend({
   validate(value) {
     if (value) {
       if (canInvoke(value, 'then')) {
-        return value.then(models => models.map(m => m.get('validations')));
+        return value.then(models => models ? models.map(m => m.get('validations')) : true);
       }
       return value.map(m => m.get('validations'));
     }

--- a/tests/dummy/app/styles/app.css
+++ b/tests/dummy/app/styles/app.css
@@ -36,10 +36,6 @@ body {
   color: #333333;
 }
 
-.demo .info img {
-  margin-bottom: 15px;
-}
-
 .demo .form {
   position: relative;
   background: #ffffff;

--- a/tests/integration/validations/model-relationships-test.js
+++ b/tests/integration/validations/model-relationships-test.js
@@ -201,6 +201,53 @@ test("belongs-to relationship is async", function(assert) {
   return validations;
 });
 
+test("belongs-to relationship is async and does not exist", function(assert) {
+  this.register('validator:belongs-to', BelongsToValidator);
+
+  var user = setupObject(this, Ember.Object.extend(BelongsToValidations), {
+    friend: new Ember.RSVP.Promise((resolve, reject) => {
+      resolve();
+    })
+  });
+
+  var validations = user.get('validations').validate();
+  assert.equal(user.get('validations.isAsync'), true);
+  assert.equal(user.get('validations.isValidating'), true);
+
+  validations.then(({
+    model, validations
+  }) => {
+    assert.equal(model, user, 'expected model to be the correct model');
+    assert.deepEqual(validations.get('content').getEach('attribute').sort(), ['friend'].sort());
+    assert.equal(user.get('validations.isValid'), true);
+  });
+
+  return validations;
+});
+
+test("has-many relationship is async and does not exist", function(assert) {
+  this.register('validator:has-many', HasManyValidator);
+
+  var user = setupObject(this, Ember.Object.extend(HasManyValidations), {
+    friends: new Ember.RSVP.Promise((resolve, reject) => {
+      resolve();
+    })
+  });
+
+  var validations = user.get('validations').validate();
+  assert.equal(user.get('validations.isAsync'), true);
+  assert.equal(user.get('validations.isValidating'), true);
+
+  validations.then(({
+    model, validations
+  }) => {
+    assert.equal(model, user, 'expected model to be the correct model');
+    assert.deepEqual(validations.get('content').getEach('attribute').sort(), ['friends'].sort());
+    assert.equal(user.get('validations.isValid'), true);
+  });
+
+  return validations;
+});
 
 test("belongs-to relationship returns undefined", function(assert) {
   this.register('validator:belongs-to', BelongsToValidator);

